### DISCUSSION TITLE
Collecting raw data for PPK using Python Interface

### DISCUSSION THREAD

**Discussion by savioratharv** [semuconsulting/PyGPSClient#153] [General]
I am currently using a u-blox ZED F9-P GNSS module and have successfully configured it to collect raw GPS data using the PyGPSClient GUI for post-processing kinematic (PPK) analysis, achieving centimeter-level accuracy. Additionally, I have a Python script that captures images from a LiDAR sensor and infrared camera at regular intervals. 

I would like to modify my current setup by integrating GPS raw data collection into this process. Specifically, I need Python code that will log raw GPS data only at the specific instances when the LiDAR and infrared cameras capture an image. The collected GPS data should be stored in a single binary log file, which can later be used for PPK processing.

**Comment by semuadmin** [comment]
@savioratharv 

> I would like to modify my current setup by integrating GPS raw data collection into this process. Specifically, I need Python code that will log raw GPS data only at the specific instances when the LiDAR and infrared cameras capture an image. The collected GPS data should be stored in a single binary log file, which can later be used for PPK processing.

OK - so are you expecting someone else to write this code for you? Binary data logging is in itself trivial, but synchronising this with an image capture routine is a very specific use case, so I doubt you'll find anything 'off the shelf'. The [FOSS GPS](https://lists.osgeo.org/listinfo/foss-gps) forum might be your best bet.

To clarify, when you say 'GPS raw data collection' I presume you mean raw *navigation* data suitable for use as a precursor to PPK processing? In the context of the ZED-F9P this typically involves enabling and capturing UBX RXM-RAWX (Multi-GNSS raw measurements) and/or RXM-SFBRX (Broadcast navigation data subframe) data and converting this to RINEX format. **Note that** PPK requires time series data so it wouldn't really make sense to capture, say, a single message corresponding to a specific camera 'shutter' event, as it were.

PyGPSClient can certainly be configured to capture binary RXM-RAWX and RXM-SFRBX data (this has been covered in your [previous thread](https://github.com/semuconsulting/PyGPSClient/discussions/147)). You could also use the [pyubx2](https://github.com/semuconsulting/pyubx2) library or [gnssstreamer](https://github.com/semuconsulting/pygnssutils?tab=readme-ov-file#gnssstreamer) CLI utility, but you would need to write your own Python script to synchronise data logging with your image capture routines - assuming these expose a suitable event API. Check out the READMEs for these libraries.

**Note that** none of these libraries currently implement RINEX conversion and there are no plans to add this functionality any time soon - you would need to use a specialised library such as [RTKLIB](https://www.rtklib.com/) to do the conversion.

**Reply by savioratharv** [reply]
Hello, I am sorry for the wording, I just wanted references or code snippets I can refer to for writing the code. Thanks for letting me know the time series aspect of it. I think my best option, in that case, would be to capture the continuous stream of UBX-RXM-RAWX and UBX-RXM-SFRBX messages as is, doing the PPK on the entire continuous data and then selecting just the points whose timestamps I am interested in. Thank you for your help.

**Reply by semuadmin** [reply]
that would be my advice - capture it all and filter it as necessary during post processing

**Reply by savioratharv** [reply]
Yes got it, thanks a lot.
